record-editor.py

```diff
@@ -265,30 +264,29 @@ def create_models( self ):
         # NOTE: we keep photo id in the model so we can pull it from our
         #       selection and access the record's data.
         #
-        self.photosModel = QStandardItemModel( 0, 3, self )
+        self.photosModel = QStandardItemModel( 0, self.NUM_COLUMNS, self )
 
-        self.photosModel.setHeaderData( 0, Qt.Horizontal, "Photo ID" )
-        self.photosModel.setHeaderData( 1, Qt.Horizontal, "File Path" )
-        self.photosModel.setHeaderData( 2, Qt.Horizontal, "State" )
+        self.photosModel.setHeaderData( self.ID_COLUMN, Qt.Horizontal, "Photo ID" )
+        self.photosModel.setHeaderData( self.PATH_COLUMN, Qt.Horizontal, "File Path" )
+        self.photosModel.setHeaderData( self.STATE_COLUMN, Qt.Horizontal, "State" )
 
         # walk through each of the photo records and insert a new item at the
         # beginning of the model's list.
-        #
-        # XXX: there has got to be a better way to insert things.  the
-        #      QAbstractItemModel doesn't have an insert that grows.
-        #      inserting QStandardItems explicitly turns our id into an
-        #      empty string.  calling reversed() on the photos does not
-        #      actually do what we want.
         for photo in self.photos:
-            self.photosModel.insertRow( 0 )
-            self.photosModel.setData( self.photosModel.index( 0, 0 ), photo["id"] )
-            self.photosModel.setData( self.photosModel.index( 0, 1 ), photo["filename"] )
-            self.photosModel.setData( self.photosModel.index( 0, 2 ), photo["state"] )
+            id_item       = QStandardItem( photo["id"] )
+            filename_item = QStandardItem( photo["filename"] )
+            state_item    = QStandardItem( photo["state"] )
+
+            id_item.setData( int( photo["id"] ) )
+            filename_item.setData( photo["filename"] )
+            state_item.setData( photo["state"] )
+
+            self.photosModel.appendRow( [id_item, filename_item, state_item] )
```

Comment:
That definitely brings it back to a sane startup time.  It is still slower
on my system but only by 5-10% (1.12s v 1.22s).  One noticeable thing is
that it's consistent in startup time while my branch has a 10% swing over
repeated runs.  Additionally, this also addresses another issue that I
didn't initially report in that clicking on the selection view's header to
sort records would cause a huge slow down as well.  Given that went away,
the startup isn't bad, and the initial sort matches the database's record
order I'll call this a win.

Definitely update this pull request with any changes made to address our
discussion.  Once everything is square I'll merge it.  Don't forget to
match the (attempted) coding style por favor.

On Sun, Aug 21, 2016 at 6:57 PM, swagneto <notifications@github.com> wrote:

> In record-editor.py
> <https://github.com/gthomsen/streetart-analysis/pull/1#discussion_r75606349>
> :
>
> >          for photo in self.photos:
> > -            self.photosModel.insertRow( 0 )
> > -            self.photosModel.setData( self.photosModel.index( 0, 0 ), photo["id"] )
> > -            self.photosModel.setData( self.photosModel.index( 0, 1 ), photo["filename"] )
> > -            self.photosModel.setData( self.photosModel.index( 0, 2 ), photo["state"] )
> > +            id_item       = QStandardItem( photo["id"] )
> > +            filename_item = QStandardItem( photo["filename"] )
> > +            state_item    = QStandardItem( photo["state"] )
> > +
> > +            id_item.setData( int( photo["id"] ) )
> > +            filename_item.setData( photo["filename"] )
> > +            state_item.setData( photo["state"] )
> > +
> > +            self.photosModel.appendRow( [id_item, filename_item, state_item] )
>
> I have an idea that may help ameliorate this. Can you apply the patch I've
> attached and let me know how it performs against your database instance? I
> was able to see a modest improvement on my build (init time went from ~0.45
> secs to ~0.3), but only against a database.xml with about 1000 photos, and
> most of which I don't have the actual photos downloaded. Also, if this does
> help performance (and you don't have an issue using a list view instead of
> a tree view), whats the best way to roll it in? Can this pull request be
> amended, or should the refactor happen in a separate request?
>
> model_patch.txt
> <https://github.com/gthomsen/streetart-analysis/files/429164/model_patch.txt>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gthomsen/streetart-analysis/pull/1/files/41ae18f71c9958a8c58e39e1eab98c472e99a783#r75606349>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJjQY1kE3XpJz8i4-RPmb6-ms4dfB14Kks5qiOX-gaJpZM4JmEfF>
> .
>